File: app/src/main/java/com/aoe/mealsapp/SettingsFragment.java
package com.aoe.mealsapp;


import android.content.Context;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.preference.PreferenceFragment;
import android.support.annotation.Nullable;
import android.util.Log;

/**
 * Allows the user to change website language (German/English) and enable/disable the
 * "Order Meal" reminder.
 */
public class SettingsFragment extends PreferenceFragment
        implements SharedPreferences.OnSharedPreferenceChangeListener {

    private static final String TAG = "SettingsFragment";

    //
    // EXTENDS PreferenceFragment
    //

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        Log.d(TAG, Thread.currentThread().getName() + " ### "
                + "onCreate() called with: savedInstanceState = [" + savedInstanceState + "]");

        addPreferencesFromResource(R.xml.preferences);

        SharedPreferences sharedPreferences = getPreferenceScreen().getSharedPreferences();

        sharedPreferences.registerOnSharedPreferenceChangeListener(this);

        /* set summary fields */

        String username = sharedPreferences.getString(SharedPreferenceKeys.USERNAME, null);
        String reminderFrequencyKey = sharedPreferences.getString(SharedPreferenceKeys.REMINDER_FREQUENCY,null);
        String languageKey = sharedPreferences.getString(SharedPreferenceKeys.LANGUAGE, null);

        // TODO handle null

        findPreference(SharedPreferenceKeys.USERNAME).setSummary(username);
        findPreference(SharedPreferenceKeys.REMINDER_FREQUENCY).setSummary(
                getActivity().getString(SharedPreferenceKeys.keysToValues.get(reminderFrequencyKey)));
        findPreference(SharedPreferenceKeys.LANGUAGE).setSummary(
                getActivity().getString(SharedPreferenceKeys.keysToValues.get(languageKey)));

    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        Log.d(TAG, Thread.currentThread().getName() + " ### "
                + "onDestroy() called");

        getPreferenceScreen().getSharedPreferences().unregisterOnSharedPreferenceChangeListener(this);
    }

    //
    // IMPLEMENTS SharedPreferences.OnSharedPreferenceChangeListener
    //

    @Override
    public void onSharedPreferenceChanged(SharedPreferences sharedPreferences, String key) {
        Log.d(TAG, Thread.currentThread().getName() + " ### "
                + "onSharedPreferenceChanged() called with: sharedPreferences = [" + sharedPreferences
                + "], key = [" + key + "]");

        switch (key) {
            case SharedPreferenceKeys.USERNAME:
                findPreference(SharedPreferenceKeys.USERNAME).setSummary(
                        sharedPreferences.getString(SharedPreferenceKeys.USERNAME, null));
                break;

            case SharedPreferenceKeys.REMINDER_FREQUENCY:
                handleReminderFrequencyChange(sharedPreferences);
                break;

            case SharedPreferenceKeys.LANGUAGE:
                handleLanguageChange(sharedPreferences);
                break;
        }
    }

    //
    // HELPER
    //

    private void handleReminderFrequencyChange(SharedPreferences sharedPreferences) {

        Context context = getActivity();

        /* set summary text in UI */

        String reminderFrequencyKey = sharedPreferences.getString(SharedPreferenceKeys.REMINDER_FREQUENCY, null); // TODO handle null
        String reminderFrequencyText = context.getString(SharedPreferenceKeys.keysToValues.get(reminderFrequencyKey));
        findPreference(SharedPreferenceKeys.REMINDER_FREQUENCY).setSummary(reminderFrequencyText);
    }

    private void handleLanguageChange(SharedPreferences sharedPreferences) {

        Context context = getActivity();

        /* set summary text in UI */

        String languageKey = sharedPreferences.getString(SharedPreferenceKeys.LANGUAGE, null); // TODO handle null
        String languageText = context.getString(SharedPreferenceKeys.keysToValues.get(languageKey));
        findPreference(SharedPreferenceKeys.LANGUAGE).setSummary(languageText);
    }
}

File: app/src/main/java/com/aoe/mealsapp/settings/SettingsFragment.kt
package com.aoe.mealsapp.settings

import android.content.Context
import android.os.Bundle
import android.preference.PreferenceFragment
import android.util.Log
import com.aoe.mealsapp.R

class SettingsFragment : PreferenceFragment(), Settings.OnSettingChangeListener {

    //
    // fragment lifecycle
    //

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        Log.d(TAG, Thread.currentThread().name + " ### "
                + "onCreate() called with: savedInstanceState = [$savedInstanceState]")

        addPreferencesFromResource(R.xml.preferences)

        Settings.getInstance(activity).registerOnSettingChangeListener(this)

        initSummaries()
    }

    override fun onDestroy() {
        super.onDestroy()
        Log.d(TAG, Thread.currentThread().name + " ### "
                + "onDestroy() called")

        Settings.getInstance(activity).unregisterOnSettingChangeListener(this)
    }

    //
    // init summaries
    //

    private fun initSummaries() {
        val settings = Settings.getInstance(activity)

        findPreference(Setting.USERNAME.toString()).summary = settings.username ?: ""

        val reminderFrequencyText = getText(activity, settings.reminderFrequency)
        findPreference(Setting.REMINDER_FREQUENCY.toString()).summary = reminderFrequencyText

        val languageText = getText(activity, settings.language)
        findPreference(Setting.LANGUAGE.toString()).summary = languageText
    }

    private fun getText(context: Context, reminderFrequency: ReminderFrequency) =
            context.getString(when (reminderFrequency) {
                ReminderFrequency.ON_SUNDAYS -> R.string.preferences_reminderFrequency_entries_onSundays
                ReminderFrequency.BEFORE_EVERY_WEEKDAY -> R.string.preferences_reminderFrequency_entries_beforeEveryWeekday
                ReminderFrequency.NEVER -> R.string.preferences_reminderFrequency_entries_never
            })

    private fun getText(context: Context, language: Language) =
            context.getString(when (language) {
                Language.ENGLISH -> R.string.preferences_language_entries_english
                Language.GERMAN -> R.string.preferences_language_entries_german
            })

    //
    // listen for settings changes
    //

    override fun onUsernameChanged(username: String) {
        findPreference(Setting.USERNAME.toString()).summary = username
    }

    override fun onPasswordChanged(password: String) {}

    override fun onLanguageChanged(language: Language) {
        val languageText = getText(activity, language)
        findPreference(Setting.LANGUAGE.toString()).summary = languageText
    }

    override fun onReminderFrequencyChanged(reminderFrequency: ReminderFrequency) {
        val reminderFrequencyText = getText(activity, reminderFrequency)
        findPreference(Setting.REMINDER_FREQUENCY.toString()).summary = reminderFrequencyText
    }

    //
    //
    //

    private companion object {
        private const val TAG = "SettingsFragment"
    }
}
